### PR TITLE
Login messages config

### PR DESCRIPTION
**Pull request by fastiuk**
This PR depends on https://github.com/fastiuk/sonic-swss-common/pull/2, https://github.com/fastiuk/sonic-host-services/pull/1
#### Why I did it
* Add an ability to configure login messages: /etc/motd, /etc/issue*
* Be able to use table defines from swsscommon

#### How to verify it
Build SONiC
or
When SONiC was already built:
```
rm -f target/python-wheels/bullseye/sonic_yang_models-1.0-py3-none-any.whl
BLDENV=bullseye make -f Makefile.work target/python-wheels/bullseye/sonic_yang_models-1.0-py3-none-any.whl
rm -f target/python-wheels/bullseye/sonic_host_services-1.0-py3-none-any.whl
BLDENV=bullseye make -f Makefile.work target/python-wheels/bullseye/sonic_host_services-1.0-py3-none-any.whl
```
Login into the switch.
Set messages:
```
redis-cli -n 4 HSET "LOGIN_MESSAGE|pre_login" "message" "Test pre login"
redis-cli -n 4 HSET "LOGIN_MESSAGE|post_login" "message" "Test post login"
```
Logout/login and see the results:
Screenshot 2022-07-02 at 22 58 11
Clear messages:
```
redis-cli -n 4 HSET "LOGIN_MESSAGE|pre_login" "message" ""
redis-cli -n 4 HSET "LOGIN_MESSAGE|post_login" "message" ""
```

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

#### Description for the changelog
* Add an ability to configure pre-login and post-login messages: /etc/motd, /etc/issue*

#### Link to config_db schema for YANG module changes
[sonic-login_message.yang](https://github.com/fastiuk/sonic-buildimage/tree/login-message-config/src/sonic-yang-models/yang-models/sonic-login_message.yang)

#### Submodules PR's :
| Repo  | PR title | State |
| ----------------- | ----------------- | ----------------- |
| sonic-host-services | [Login messages config](https://github.com/fastiuk/sonic-host-services/pull/1) | GitHub issue/pull request detail |
| sonic-swss-common | [Login messages config](https://github.com/fastiuk/sonic-swss-common/pull/2) | GitHub issue/pull request detail |

### CODE REVIEW

**Comment by fastiuk** [comment]
@stepanblyschak please review

**Comment by stepanblyschak** [comment]
@fastiuk What is the use case?

**Comment by fastiuk** [comment]
> @fastiuk What is the use case?

The use case is to customize/rebrand your switch.
For example: here is a default Login screen:
Screenshot 2022-07-04 at 18 05 24
After customizing it will look like this (or in any other way you want):
Screenshot 2022-07-04 at 18 33 51

**Comment by stepanblyschak** [comment]
@fastiuk What is the use case for a user to customize this message through config db?
If you want to create an NVIDIA distribution of SONiC you can modify this file at image build time.

**Comment by fastiuk** [comment]
> What is the use case for a user to customize this message through config db?
> If you want to create an NVIDIA distribution of SONiC you can modify this file at image build time.

The use case is to be able to configure it at runtime. It is not a good user experience if it needs to rebuild an image in order to change the login messages. From the user's perspective of view: the user will grab the prebuilt image from github/artifactory/etc install it and then configure it out-of-box without a need to rebuild it.

File: src/sonic-host-services/scripts/hostcfgd [hunk 1]
```diff
@@ -1253,6 +1254,69 @@ class PamLimitsCfg(object):
...
+        login_message_pre = '''sudo bash -c "echo -e{} '{}' > /etc/issue.net"'''
+        login_message_appl = "sudo sed -iE 's/\(#\|\)Banner.*$/Banner "\
+                            "\/etc\/issue.net/' /etc/ssh/sshd_config"
+        login_message_rest_ssh = "sudo service ssh restart"
+        login_message_post = '''sudo bash -c "echo -e{} '{}' > /etc/motd"'''
+
+    def __init__(self):
+        self.cache = {}
+
+    def login_message(self, key, data):
```
Comment (stepanblyschak) [inline src/sonic-host-services/scripts/hostcfgd]:
```suggestion
    def set_login_message(self, key, data):
```

File: src/sonic-host-services/scripts/hostcfgd [hunk 2]
```diff
@@ -1253,6 +1254,69 @@ class PamLimitsCfg(object):
...
+            db: DB instance.
+            table: DB table that was changed.
+            key: DB table's key that was triggered change.
+            data: Read table data.
+        """
+        # Handling pre/post login messages. Data should be a dict
+        msg = None if not isinstance(data, dict) else data.get('message')
+        if type(msg) != str:
+            # Nothing to handle
+            return
```
Comment (stepanblyschak) [inline src/sonic-host-services/scripts/hostcfgd]: Why type checks are needed ? ```isinstance``` and ```type```?

File: src/sonic-host-services/scripts/hostcfgd [hunk 3]
```diff
@@ -1253,6 +1254,69 @@ class PamLimitsCfg(object):
...
+
+        # Check with cache
+        if msg == self.cache.get(key):
+            # Nothing to handle
+            return
```
Comment (stepanblyschak) [inline src/sonic-host-services/scripts/hostcfgd]: Is maintaining a cache really needed? Is the performance gains over cost of maintaining a valid cache in the code justified?

File: src/sonic-host-services/scripts/hostcfgd [hunk 4]
```diff
@@ -1253,6 +1254,69 @@ class PamLimitsCfg(object):
...
+
+        # Don't print newline when message is empty
+        not_newline = '' if msg else 'n'
+
+        try:
+            # Check which key was changed
+            if key == 'pre_login':
+                # Write pre-login to /etc/issue file
+                run_cmd(self.cmds.login_message_pre.format(not_newline, msg),
+                        True, True)
```
Comment (stepanblyschak) [inline src/sonic-host-services/scripts/hostcfgd]: use explicit keyword arguments, like login_err=True to improve readability

File: src/sonic-host-services/scripts/hostcfgd [hunk 5]
```diff
@@ -1253,6 +1254,69 @@ class PamLimitsCfg(object):
...
+            elif key == 'post_login':
+                # Write post-login to /etc/motd file
+                run_cmd(self.cmds.login_message_post.format(not_newline, msg),
+                        True, True)
+        except Exception:
+            syslog.syslog(syslog.LOG_ERR, f'Failed to update {key} message')
+            return
+
+        # Update cache
+        syslog.syslog(f'Set new {key} message: {msg}')
```
Comment (stepanblyschak) [inline src/sonic-host-services/scripts/hostcfgd]: What is the severity of this message? INFO, DEBUG?

File: src/sonic-yang-models/tests/files/sample_config_db.json [hunk 6]
```diff
@@ -1879,25 +1879,26 @@
...
-             "CUSTOM_L3": {
-                 "matches": [
-                     "IN_PORTS",
-                     "OUT_PORTS",
-                     "SRC_IP"
-                 ],
-                 "actions": [
-                     "PACKET_ACTION",
-                     "MIRROR_INGRESS_ACTION"
-                 ],
```
Comment (stepanblyschak) [inline src/sonic-yang-models/tests/files/sample_config_db.json]: Don't change these lines

File: src/sonic-yang-models/tests/files/sample_config_db.json [hunk 7]
```diff
@@ -1911,6 +1912,15 @@
                 "default_dynamic_th": "0",
                 "over_subscribe_ratio": "0"
             }
+        },
+
```
Comment (stepanblyschak) [inline src/sonic-yang-models/tests/files/sample_config_db.json]: Remove empty line

File: src/sonic-yang-models/yang-models/sonic-login_message.yang [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+                    type string;
+                }
+
+            }
+        }
+        /* end of container LOGIN_MESSAGE */
+    }
+    /* end of top level container */
+}
+/* end of module nvue */
```
Comment (stepanblyschak) [inline src/sonic-yang-models/yang-models/sonic-login_message.yang]: End of module sonic-login_message

**Review by stepanblyschak — CHANGES_REQUESTED** [review]
Do not include submodule update in this PR. Submodule update should be done in a separate PR.

I understand the motivation to change the "welcome" message for custom images. AFAIK Alibaba, Kuaishu, etc. - those companies build their own distribution based on SONiC and they change the welcome message. That is nice, because when you enter the switch you know is SONiC based you see which company's distribution you are on as every company has own modifications to the image.

However, I do not understand the benefit of having this configurable though CONFIG DB and letting the user to configure it. What does user gain by doing so? This seems purelly cosmetic and can be done by editing files mannually by the user since SONiC gives you full access to the  filesystem. We will have to come up with an HLD with this proposal and review with community first before internal code review.

File: src/sonic-host-services/scripts/hostcfgd (hunk 5 above)
Comment (fastiuk) [reply src/sonic-host-services/scripts/hostcfgd]: default for syslog is INFO

File: src/sonic-host-services/scripts/hostcfgd (hunk 3 above)
Comment (fastiuk) [reply src/sonic-host-services/scripts/hostcfgd]: I did the same as other "Cfg" classes do.

File: src/sonic-host-services/scripts/hostcfgd (hunk 2 above)
Comment (fastiuk) [reply src/sonic-host-services/scripts/hostcfgd]: to avoid unexpected fails.
If something bad will come from outside we will filter it here.

**Comment by fastiuk** [comment]
Hostcfgd moved here: https://github.com/fastiuk/sonic-host-services/pull/1

**Comment by fastiuk** [comment]
Hey @stepanblyschak . Can we create community PR for this?

**Review by stepanblyschak — APPROVED** [review]

**Pull request closed without merging**